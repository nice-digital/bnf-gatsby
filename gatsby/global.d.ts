// Let TypeScript know we're importing SCSS (and SCSS modules)
// Avoids "Can't import CSS/SCSS modules. TypeScript says “Cannot Find Module”"
// See https://stackoverflow.com/a/41946697/486434
declare module "*.module.scss" {
	const content: { [className: string]: string };
	export default content;
}

// So that we can import JSON files in TypeScript.
declare module "*.json" {
	const value: unknown;

	export default value;
}
