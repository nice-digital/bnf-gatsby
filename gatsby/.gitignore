node_modules/
.cache/
public

# dotenv environment variables file
.env
.env.*
