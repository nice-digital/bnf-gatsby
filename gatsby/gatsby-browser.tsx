import {
	ShouldUpdateScrollArgs,
	type RouteUpdateArgs,
	type WrapPageElementBrowserArgs,
} from "gatsby";
import { type ReactElement } from "react";

import { Layout } from "@/components/Layout/Layout";

// Gatsby hook for when the route has changed on the client side
// See https://www.gatsbyjs.org/docs/browser-apis/#onRouteUpdate
export const onRouteUpdate = ({
	prevLocation,
	location,
}: RouteUpdateArgs): void => {
	if (prevLocation) {
		// Cast is needed because of https://github.com/gatsbyjs/gatsby/issues/29124
		const prevPath = (prevLocation as Location).pathname,
			path = location.pathname;

		// Push our own event to the dataLayer on page change rather than using the
		// 'gatsby-route-change' event built into gatsby-plugin-google-tagmanager.
		// Because gatsby-route-change is pushed on initial page load as well as route change,
		// AND because it doesn't use requestAnimationFrame to delay until the page title has been updated

		const sendPageView = () => {
			window.dataLayer.push({
				location: location.href,
				referrer: prevLocation.href,
			});
			// Don't consider hash changes to be a page view - pageviews only happy when the path changes
			if (prevPath != path) window.dataLayer.push({ event: "pageview" });
		};

		// Delay before push to the data layer, to make sure the page title has been updated
		// See https://github.com/gatsbyjs/gatsby/pull/10917/files#diff-bf0d94c8bf47d5c1687e342c2dba1e00R12-R13
		if ("requestAnimationFrame" in window) {
			window.requestAnimationFrame(() => {
				window.requestAnimationFrame(sendPageView);
			});
		} else {
			// simulate 2 rAF calls
			setTimeout(sendPageView, 32);
		}
	} else {
		window.dataLayer.push({
			location: location.href,
			referrer: document.referrer,
		});
	}
	// Default return statement when prevLocation is falsy
	return;
};

export const wrapPageElement = ({
	element,
	props,
}: WrapPageElementBrowserArgs): ReactElement => (
	<Layout {...props}>{element}</Layout>
);

/**
 * Gatsby hook for overriding scroll position
 * See https://www.gatsbyjs.org/docs/browser-apis/#shouldUpdateScroll
 */
export const shouldUpdateScroll = ({
	routerProps: { location },
	prevRouterProps,
	getSavedScrollPosition,
}: ShouldUpdateScrollArgs): boolean | string | [number, number] => {
	if (
		// If there's no previous route props we're coming from an external site, which means
		// we want to scroll to the hash (if there is one), and _not_ a saved scroll position
		!prevRouterProps ||
		// Or we're linking to as hash within the same page
		(prevRouterProps.location.pathname === location.pathname && location.hash)
	) {
		// Provide our own scroll to hash to avoid Gatsby using a stored scroll position
		requestAnimationFrame(() => {
			requestAnimationFrame(() => {
				const targetElement = location.hash
					? document.querySelector(location.hash)
					: null;

				if (targetElement) {
					targetElement.setAttribute("tabIndex", "-1");
					(targetElement as HTMLElement).focus();
					targetElement.scrollIntoView();
				}
			});
		});
		return false;
	}

	const savedScrollY = (getSavedScrollPosition(location)?.[1] || 0) as number;
	if (savedScrollY > 0) {
		window.scrollTo(0, savedScrollY);
		return false;
	}

	if (location.hash) {
		requestAnimationFrame(() => {
			requestAnimationFrame(() => {
				const targetElement = document.querySelector(location.hash);

				// Default to Gatsby's default behaviour if the element doesn't exist
				if (!targetElement) return true;

				// Provide our own scroll to hash to avoid Gatsby using a stored scroll position
				targetElement.setAttribute("tabIndex", "-1");
				(targetElement as HTMLElement).focus();
				targetElement.scrollIntoView();
				return false;
			});
		});
		return false;
	}

	requestAnimationFrame(() => {
		requestAnimationFrame(() => {
			const contentStartElement = document.getElementById("content-start");

			if (!contentStartElement) return true;

			contentStartElement.setAttribute("tabIndex", "-1");

			//scrolling to 0,0 to preserve existing BNF navigation behaviour
			window.scrollTo(0, 0);
			return false;
		});
	});
	// Default to scrolling to the content start element as the standard navigation behaviour

	return true;
};
