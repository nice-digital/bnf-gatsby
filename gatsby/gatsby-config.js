require("dotenv").config({
	path: `.env.${process.env.NODE_ENV}`,
});

const isBNF = process.env.GATSBY_SITE === "bnf";

if (!process.env.GATSBY_SEARCH_URL)
	throw new Error(
		"Env var GATSBY_SEARCH_URL isn't set. Did you forget to add it?"
	);

module.exports = {
	jsxRuntime: "automatic",
	siteMetadata: {
		isBNF,
		siteUrl: `https://${isBNF ? "bnf" : "bnfc"}.nice.org.uk`,
		siteTitleShort: isBNF ? "BNF" : "BNFC",
		siteTitleLong:
			"British National Formulary" + (isBNF ? "" : " for Children"),
	},
	plugins: [
		{
			resolve: `gatsby-source-bnf`,
			options: {
				feedURL: process.env.FEED_URL,
				userKey: process.env.FEED_USER_KEY,
				site: isBNF ? "bnf" : "bnfc",
			},
		},
		// Avoid errors like "ModuleNotFoundError: Module not found: Error: Can't resolve '@/components/Layout/Layout'" when using custom paths in tsconfig.json
		`gatsby-plugin-tsconfig-paths`,
		{
			resolve: `gatsby-plugin-sass`,
			options: {
				// Make CSS modules play nicely with TypeScript, see also global.d.ts for *.scss TS def
				cssLoaderOptions: {
					esModule: false,
					modules: {
						namedExport: false,
					},
				},
			},
		},
		"gatsby-plugin-catch-links",
		"gatsby-plugin-react-helmet",
		"gatsby-plugin-remove-generator",
		{
			resolve: `gatsby-plugin-sitemap`,
			options: {
				excludes: [`/search/`],
			},
		},
		{
			resolve: "gatsby-plugin-google-tagmanager",
			options: {
				id: "GTM-5H5L9GK",
				includeInDevelopment: true,
			},
		},
		{
			resolve: "gatsby-plugin-manifest",
			options: {
				icon: "src/images/icon.png",
			},
		},
	],
};
