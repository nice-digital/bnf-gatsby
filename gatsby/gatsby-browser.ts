import { RouteUpdateArgs } from "gatsby";

// Gatsby hook for when the route has changed on the client side
// See https://www.gatsbyjs.org/docs/browser-apis/#onRouteUpdate
export const onRouteUpdate = ({
	prevLocation,
	location,
}: RouteUpdateArgs): void => {
	if (prevLocation) {
		// Cast is needed because of https://github.com/gatsbyjs/gatsby/issues/29124
		const prevPath = (prevLocation as Location).pathname,
			path = location.pathname;

		// Push our own event to the dataLayer on page change rather than using the
		// 'gatsby-route-change' event built into gatsby-plugin-google-tagmanager.
		// Because gatsby-route-change is pushed on initial page load as well as route change,
		// AND because it doesn't use requestAnimationFrame to delay until the page title has been updated

		const sendPageView = () => {
			window.dataLayer.push({
				location: location.href,
				referrer: prevLocation.href,
			});
			// Don't consider hash changes to be a page view - pageviews only happy when the path changes
			if (prevPath != path) window.dataLayer.push({ event: "pageview" });
		};

		// Delay before push to the data layer, to make sure the page title has been updated
		// See https://github.com/gatsbyjs/gatsby/pull/10917/files#diff-bf0d94c8bf47d5c1687e342c2dba1e00R12-R13
		if ("requestAnimationFrame" in window) {
			window.requestAnimationFrame(() => {
				window.requestAnimationFrame(sendPageView);
			});
		} else {
			// simulate 2 rAF calls
			setTimeout(sendPageView, 32);
		}
	} else {
		window.dataLayer.push({
			location: location.href,
			referrer: document.referrer,
		});
	}
};
