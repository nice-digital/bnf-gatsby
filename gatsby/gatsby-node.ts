import { CreateSchemaCustomizationArgs, CreateWebpackConfigArgs } from "gatsby";

import { initialFieldExtension } from "./src/field-extensions/initial";

/**
 * Gatsby hook for customizing the schema.
 * See https://www.gatsbyjs.org/docs/schema-customization/
 */
export const createSchemaCustomization = ({
	actions: { createFieldExtension, createTypes },
}: CreateSchemaCustomizationArgs): void => {
	createFieldExtension(initialFieldExtension);

	// Add initials for types used in A-Z lists, as it's easier for grouping
	// We don't add the initial field in the source plugin because grouping is a presentation concern
	const typeDefs = `
	  type BnfDrug implements Node {
		"The lowercase first letter of the title, used for grouping"
		initial: String! @initial(field: "title")
	  }
	  type BnfTreatmentSummary implements Node {
		"The lowercase first letter of the title, used for grouping"
		initial: String! @initial(field: "title")
	  }
	  type BnfInteractant implements Node {
		"The lowercase first letter of the title, used for grouping"
		initial: String! @initial(field: "title")
	  }
	`;
	createTypes(typeDefs);
};

/**
 * Gatsby hook for overriding webpack config
 * See https://www.gatsbyjs.com/docs/how-to/custom-configuration/add-custom-webpack-config/
 */
export const onCreateWebpackConfig = ({
	stage,
	actions,
	getConfig,
	plugins,
}: CreateWebpackConfigArgs): void => {
	// Disable order warnings from mini CSS plugin
	// See https://github.com/gatsbyjs/gatsby/discussions/30169#discussioncomment-621285
	if (stage === "build-javascript" || stage === "develop") {
		const config = getConfig();

		const miniCss = config.plugins.find(
			(plugin: () => void) => plugin.constructor.name === "MiniCssExtractPlugin"
		);

		if (miniCss) {
			miniCss.options.ignoreOrder = true;
		}

		actions.replaceWebpackConfig(config);

		actions.setWebpackConfig({
			plugins: [plugins.provide({ process: "process/browser" })],
		});
	}
};
