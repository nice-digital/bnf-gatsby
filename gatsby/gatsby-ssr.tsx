import { PreRenderHTMLArgs } from "gatsby";
import React from "react";

export const onPreRenderHTML = ({
	getHeadComponents,
	replaceHeadComponents,
}: PreRenderHTMLArgs): void => {
	const components = getHeadComponents();

	components.push(
		<script
			src={process.env.GATSBY_COOKIE_BANNER_URL}
			type="text/javascript"
			key="cookie-banner"
			async
		></script>
	);

	replaceHeadComponents(components);
};
