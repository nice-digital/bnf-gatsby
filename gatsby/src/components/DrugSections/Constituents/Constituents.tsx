import { Link } from "gatsby";
import { type FC } from "react";

import { isTruthy, type SlugAndTitle } from "@/utils";

import styles from "../DrugSection.module.scss";
import { BasePot } from "../types";

export type ConstituentsProps = BasePot & {
	message: string;
	constituents: SlugAndTitle[];
};

export const Constituents: FC<ConstituentsProps> = ({
	message,
	constituents,
	potName,
	slug,
}) => {
	return (
		<section aria-labelledby={slug} className={styles.section}>
			<h2 id={slug} dangerouslySetInnerHTML={{ __html: potName }} />
			<p dangerouslySetInnerHTML={{ __html: message }} />
			<ul aria-labelledby={slug}>
				{constituents.filter(isTruthy).map(({ slug, title }) => (
					<li key={slug}>
						<Link
							to={`/drugs/${slug}/`}
							dangerouslySetInnerHTML={{ __html: title }}
						/>
					</li>
				))}
			</ul>
		</section>
	);
};
