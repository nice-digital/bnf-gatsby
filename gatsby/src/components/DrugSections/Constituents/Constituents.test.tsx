import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { navigate } from "gatsby";

import { Constituents, type ConstituentsProps } from "./Constituents";

describe("Constituents", () => {
	const props: ConstituentsProps = {
		potName: "Constituent drugs",
		slug: "constituent-drugs",
		message:
			"The properties listed <em>below</em> are those particular to the combination only",
		constituents: [
			{
				title: "bimatoprost",
				slug: "bimatoprost",
			},
			{
				title: "<em>timolol</em> maleate",
				slug: "timolol-maleate",
			},
		],
	};

	it("should render section with accessible name", () => {
		render(<Constituents {...props} />);

		expect(
			screen.getByRole("region", { name: "Constituent drugs" })
		).toBeInTheDocument();
	});

	it("should render heading 2 with section name", () => {
		render(<Constituents {...props} />);

		expect(screen.getByRole("heading", { level: 2 })).toHaveTextContent(
			"Constituent drugs"
		);
	});

	it("should render message as HTML", () => {
		render(<Constituents {...props} />);

		expect(
			screen.getByText(
				(_content, node) =>
					node?.textContent ===
					"The properties listed below are those particular to the combination only",
				{ selector: "p" }
			)
		).toBeInTheDocument();
	});

	it("should list with accessible name", () => {
		render(<Constituents {...props} />);

		expect(
			screen.getByRole("list", { name: "Constituent drugs" })
		).toBeInTheDocument();
	});

	it("should render link to drug monograph page for each constituent", () => {
		render(<Constituents {...props} />);

		expect(screen.getAllByRole("link")).toHaveLength(2);
	});

	it("should render drug HTML name as link text", () => {
		render(<Constituents {...props} />);

		expect(
			screen.getByRole("link", { name: "timolol maleate" })
		).toBeInTheDocument();
	});

	it("should link to drug monograph page", () => {
		render(<Constituents {...props} />);

		expect(
			screen.getByRole("link", { name: "timolol maleate" })
		).toHaveAttribute("href", "/drugs/timolol-maleate/");
	});

	it("should use Gatsbys links component for mongraph links", async () => {
		render(<Constituents {...props} />);

		userEvent.click(screen.getByRole("link", { name: "bimatoprost" }));
		await waitFor(() => {
			expect(navigate).toHaveBeenCalledWith("/drugs/bimatoprost/");
		});
	});

	it("should match snapshot", () => {
		render(<Constituents {...props} />);

		expect(
			screen.getByRole("region", { name: "Constituent drugs" })
		).toMatchSnapshot();
	});
});
