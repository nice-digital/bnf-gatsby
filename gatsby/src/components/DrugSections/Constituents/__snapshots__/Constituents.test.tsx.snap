// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Constituents should match snapshot 1`] = `
<section
  aria-labelledby="constituent-drugs"
  class="section"
>
  <h2
    id="constituent-drugs"
  >
    Constituent drugs
  </h2>
  <p>
    The properties listed 
    <em>
      below
    </em>
     are those particular to the combination only
  </p>
  <ul
    aria-labelledby="constituent-drugs"
  >
    <li>
      <a
        href="/drugs/bimatoprost/"
      >
        bimatoprost
      </a>
    </li>
    <li>
      <a
        href="/drugs/timolol-maleate/"
      >
        <em>
          timolol
        </em>
         maleate
      </a>
    </li>
  </ul>
</section>
`;
